api/types/provisioning.go

```diff
@@ -575,26 +583,49 @@ func (a *ProvisionTokenSpecV2CircleCI) checkAndSetDefaults() error {
 
 func (a *ProvisionTokenSpecV2Kubernetes) checkAndSetDefaults() error {
 	if len(a.Allow) == 0 {
-		return trace.BadParameter(
-			"the %q join method requires defined kubernetes allow rules",
-			JoinMethodKubernetes,
-		)
+		return trace.BadParameter("allow: at least one rule must be set")
 	}
-	for _, allowRule := range a.Allow {
+	for i, allowRule := range a.Allow {
 		if allowRule.ServiceAccount == "" {
 			return trace.BadParameter(
-				"the %q join method requires kubernetes allow rules with non-empty service account name",
-				JoinMethodKubernetes,
+				"allow[%d].service_account: name of service account must be set",
+				i,
 			)
 		}
 		if len(strings.Split(allowRule.ServiceAccount, ":")) != 2 {
 			return trace.BadParameter(
-				`the %q join method service account rule format is "namespace:service_account", got %q instead`,
-				JoinMethodKubernetes,
+				`allow[%d].service_account: name of service account should be in format "namespace:service_account", got %q instead`,
+				i,
 				allowRule.ServiceAccount,
 			)
 		}
 	}
+
+	if a.Type == KubernetesJoinTypeUnspecified {
+		// For compatibility with older resources which did not have a Type
+		// field we default to "in_cluster".
+		a.Type = KubernetesJoinTypeInCluster
+	}
+	switch a.Type {
+	case KubernetesJoinTypeInCluster:
+		if a.StaticJWKS != nil {
+			return trace.BadParameter("static_jwks: must not be set when type is %q", KubernetesJoinTypeInCluster)
+		}
+	case KubernetesJoinTypeStaticJWKS:
+		if a.StaticJWKS == nil {
+			return trace.BadParameter("static_jwks: must be set when type is %q", KubernetesJoinTypeStaticJWKS)
+		}
+		if a.StaticJWKS.JWKS == "" {
+			return trace.BadParameter("static_jwks.jwks: must be set when type is %q", KubernetesJoinTypeStaticJWKS)
+		}
```

Comment:
I nearly added this - but I wasn't sure if we'd want to add the jwt package as a dependency of the `api` module. Open to thoughts on this, it'd be relatively easy to add but it seems like a big dependency to add to validate a single field on a single type.

Reply:
Oh, right this would inflate the client binaries. I think I'll get skinned alive by the scale team if I add 5 MiB to `tsh` again  😅 .

Can we check this server-side on resource creation/update? If it takes more than 5 minutes don't bother, this is a nit anyway.

Reply:
I think it'd be a little weird to work into the creation/update - but I'll take a look and make the change if it's not too weird.